src/lib/sql/sql_pipeline.cpp

```diff
@@ -1,163 +1,158 @@
 #include "optimizer/optimizer.hpp"
 #include "scheduler/current_scheduler.hpp"
-#include "sql_pipeline.hpp"
-#include "sql_translator.hpp"
+#include "sql/sql_translator.hpp"
 #include "utils/assert.hpp"
 
 namespace opossum {
 
-SQLPipeline::SQLPipeline(const std::string& sql, std::shared_ptr<TransactionContext> transaction_context)
-    : _sql_string(sql), _use_mvcc(true), _auto_commit(false), _transaction_context(std::move(transaction_context)) {
-  DebugAssert(_transaction_context != nullptr, "Cannot pass nullptr as explicit transaction context.");
+std::vector<SQLPipeline> SQLPipeline::from_sql_string(const std::string& sql, const bool use_mvcc) {
+  // Pass no explicit transaction context so the SQLPipeline can decide whether it should create one or not.
+  return _from_sql_string(sql, nullptr, use_mvcc);
 }
 
-SQLPipeline::SQLPipeline(const std::string& sql, const bool use_mvcc)
-    : _sql_string(sql), _use_mvcc(use_mvcc), _auto_commit(_use_mvcc) {
-  if (_use_mvcc) {
-    // We want to use MVCC but didn't pass an explicit context
-    _transaction_context = TransactionManager::get().new_transaction_context();
-  } else {
-    // We don't want to use MVCC
-    _transaction_context = nullptr;
-  }
+std::vector<SQLPipeline> SQLPipeline::from_sql_string(const std::string& sql,
+                                                      std::shared_ptr<TransactionContext> transaction_context) {
+  DebugAssert(transaction_context != nullptr, "Cannot pass nullptr as explicit transaction context.");
+  return _from_sql_string(sql, transaction_context, true);
 }
 
-const hsql::SQLParserResult& SQLPipeline::get_parsed_sql() {
-  if (_parsed_sql) {
-    // Returned cached result
-    return *(_parsed_sql.get());
-  }
-
-  auto parse_result = std::make_unique<hsql::SQLParserResult>();
-
-  const auto started = std::chrono::high_resolution_clock::now();
+std::vector<SQLPipeline> SQLPipeline::_from_sql_string(const std::string& sql,
+                                                       std::shared_ptr<TransactionContext> transaction_context,
+                                                       bool use_mvcc) {
+  hsql::SQLParserResult parse_result;
   try {
-    hsql::SQLParser::parse(_sql_string, parse_result.get());
+    hsql::SQLParser::parse(sql, &parse_result);
   } catch (const std::exception& exception) {
     throw std::runtime_error("Error while parsing SQL query:\n  " + std::string(exception.what()));
   }
 
-  if (!parse_result->isValid()) {
-    throw std::logic_error("SQL query not valid.");
+  if (!parse_result.isValid()) {
+    throw std::runtime_error(_create_parse_error_message(sql, parse_result));
   }
 
-  const auto done = std::chrono::high_resolution_clock::now();
-  _parse_time_sec = std::chrono::duration_cast<std::chrono::microseconds>(done - started);
+  auto pipelines = std::vector<SQLPipeline>();
+  pipelines.reserve(parse_result.size());
+
+  for (auto* statement : parse_result.releaseStatements()) {
```

Comment:
What happens if this fails somewhere? Do the remaining results get released or do they leak?